update-controlled-shapes.py

```diff
@@ -0,0 +1,84 @@
+import logging
+import typer
+from rdflib.namespace import RDF, RDFS, VOID, XSD, SOSA, DCTERMS, PROV, TIME, SH
+from rdflib import Graph, Namespace, URIRef, Literal, BNode
+
+app = typer.Typer()
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+@app.command()
+def fetch(filename: str):
+
+    g = Graph()
+
+    g.parse(filename)
+
+    q_delete = """
+        PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+        PREFIX sh: <http://www.w3.org/ns/shacl#>
+        delete {
+            ?uri sh:property ?property .
+            ?property sh:in ?node .
+            ?node rdf:rest ?item .
+            ?item rdf:first ?head ;
+                rdf:rest ?tail .
+            ?property sh:path ?path .
+        }
+        where {
+            ?uri sh:property ?property .
+            ?property sh:in ?node .
+            ?node rdf:rest* ?item .
+            ?item rdf:first ?head ;
+                rdf:rest ?tail .
+            ?property sh:path ?path .
+        }
+    """
+
+    logger.info(f"Delete sh:in in file '{filename}'")
+    g.update(q_delete)
+
+    q = """
+        select ?uri ?query {
+            ?uri a <urn:class:Controlled> ;
+            <urn:property:query> ?query .
+        }
+    """
+
+    for query in g.query(q).bindings:
+        start_from_the_last_blank_node = True
+        uri = query["uri"]
+        logger.info(f"Updating sh:in of shape '{uri}' in '{filename}'")
+        node_property = BNode()
+        for r in g.query(query["query"]):
+            if start_from_the_last_blank_node:
+                last_node = BNode()
+                g.add((last_node, RDF.first, r["values"]))
+                g.add((last_node, RDF.rest, RDF.nil))
+                start_from_the_last_blank_node = False
+            else:
+                next_node = BNode()
+                g.add((next_node, RDF.first, r["values"]))
+                g.add((next_node, RDF.rest, last_node))
+                last_node = next_node
+
+        g.add((node_property, SH.path, RDF.value))
+        g.add((node_property, SH["in"], last_node))
+
+        g.add(
+            (
+                uri,
+                SH.property,
+                node_property,
+            )
+        )
+
+        logger.info(f"Updated sh:in of shape '{uri}' in '{filename}'")
```

Comment:
- loop through the file, get a list of controlled shapes
- for each shape, fetch the vocabs from the sparql query with proper exception handling
- if successful, delete the items in sh:in and add the new items, write to file
- if not successful, log an error and don't delete anything